src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp

```diff
@@ -472,11 +472,73 @@ void PDPClient::assignRemoteEndpoints(
         }
     }
 
-    notifyAboveRemoteEndpoints(*pdata);
+#if HAVE_SECURITY
+    if (mp_RTPSParticipant->security_manager().discovered_participant(*pdata))
+#endif // HAVE_SECURITY
+    {
+        perform_builtin_endpoints_matching(*pdata);
+    }
 }
 
 void PDPClient::notifyAboveRemoteEndpoints(
         const ParticipantProxyData& pdata)
+{
+#if HAVE_SECURITY
+    if (should_protect_discovery())
+    {
+        eprosima::shared_lock<eprosima::shared_mutex> disc_lock(mp_builtin->getDiscoveryMutex());
+
+        // Verify if this participant is a server
+        for (auto& svr : mp_builtin->m_DiscoveryServers)
+        {
+            if (svr.guidPrefix == pdata.m_guid.guidPrefix)
+            {
+                match_pdp_reader_nts_(svr, pdata.m_guid.guidPrefix);
+                match_pdp_writer_nts_(svr, pdata.m_guid.guidPrefix);
+                break;
+            }
+        }
+    }
+#endif // HAVE_SECURITY
+
+    perform_builtin_endpoints_matching(pdata);
+}
+
```

Comment:
I dont know if this should be within an `else {}` statement. If HAVE_SECURITY is true and `should_protect_discovery()` returns true, perhaps it could be called twice, one here and again in assignRemoteEndpoints().

```suggestion
        }
    }
    else
#endif // HAVE_SECURITY
    {
        perform_builtin_endpoints_matching(pdata);
    }
}

```

Reply:
Nice catch. But I think the correct solution would be to remove the call to `perform_builtin_endpoints_matching` from here.

Reply:
Okey. Although having a second glance at it perhaps it is not a bad idea to leave it as it is, so we ensure we call it. 
The `perform_builtin_endpoints_matching()` call within `assignRemoteEndpoints()` is not always performed in every callback invokation because the `if(mp_RTPSParticipant->security_manager().discovered_participant(*pdata))` can return false (despite participant already exists) due to the fact that AuthInfoUniquePtr in the discovered_participants_ collection can be busy for that participant (other thread may have taken the ownership as the pointer is being moved from the collection in a lot of methods).

BTW, I have the mangled guid PR ready, will upload later after local testing.